clients/esp32/client.ino

```diff
@@ -0,0 +1,272 @@
+
+    const size_t numSamples = rawData.size() / sizeof(int16_t);
+    std::vector<int32_t> scaledAudioData(numSamples);
+
+    // This initial scaling factor moves int16 range to int32
+    int32_t initialScalingFactor = 1 << 16;
+
+    for(size_t i = 0; i < numSamples; i++) {
+        int16_t sample16 = ((int16_t*)rawData.data())[i]; 
+        // Convert to int32 and apply initial scaling to use the full range of int32
+        int32_t sample32 = ((int32_t)sample16) * initialScalingFactor;
+        scaledAudioData[i] = sample32;
+    }
+
+    size_t bytesWritten;
+    esp_err_t err = i2s_write(I2S_NUM_0, scaledAudioData.data(), scaledAudioData.size() * sizeof(int32_t), &bytesWritten, portMAX_DELAY);
+}
+
+void onEventsCallback(WebsocketsEvent event, String data) {
+    if(event == WebsocketsEvent::ConnectionOpened) {
+        Serial.println("Connnection Opened");
+    } else if(event == WebsocketsEvent::ConnectionClosed) {
+        Serial.println("Connnection Closed");
+    } else if(event == WebsocketsEvent::GotPing) {
+        Serial.println("Got a Ping!");
+    } else if(event == WebsocketsEvent::GotPong) {
+        Serial.println("Got a Pong!");
+    }
+}
+
+void connectToWifi(String ssid, String password) {
+    WiFi.begin(ssid.c_str(), password.c_str());
+
+    // Wait for connection to establish
+    int attempts = 0;
+    while (WiFi.status() != WL_CONNECTED && attempts < 20) {
+        delay(1000);
+        Serial.print(".");
+        attempts++;
+    }
+
+    if(WiFi.status() == WL_CONNECTED) {
+        Serial.println("Connected to Wi-Fi");
+    } else {
+        Serial.println("Failed to connect to Wi-Fi. Check credentials.");
+    }
+}
+
+void setup() {
+  // put your setup code here, to run once:
+  Serial.begin(115200);
+  Serial.println("setting up wifi");
+  WiFi.mode(WIFI_STA);
+  connectToWifi(ssid, password);
+  Serial.println("connected to wifi");
+  Serial.println("let's try playing some audio");
+
+  i2s_config_t i2s_config = {
+    .mode = (i2s_mode_t)(I2S_MODE_MASTER | I2S_MODE_RX | I2S_MODE_TX),
+    .sample_rate = 24000, // Adjust this according to your audio file's sample rate
+    .bits_per_sample = I2S_BITS_PER_SAMPLE_32BIT, // Adjust this too
+    .channel_format = I2S_CHANNEL_FMT_ONLY_LEFT, // Mono
+    .communication_format = I2S_COMM_FORMAT_I2S,
+    .intr_alloc_flags = 0, // Default interrupt priority
+    .dma_buf_count = 4,
+    .dma_buf_len = 1024,
+    .use_apll = false,
+    .tx_desc_auto_clear = true, // Auto clear tx descriptor on underflow
+    .fixed_mclk = 0
+  };
+
+  i2s_pin_config_t pin_config = {
+    .bck_io_num = 16, // BCLK
+    .ws_io_num = 9, // LRCK
+    .data_out_num = 15, // Speaker output
+    .data_in_num = 17 // Microphone input
+  };
+
+  // Install and start I2S driver with your new pin configuration
+  i2s_driver_install(i2s_num, &i2s_config, 0, NULL);
+  i2s_set_pin(i2s_num, &pin_config);  
+
+  i2s_set_sample_rates(I2S_NUM_0, 22050); // Set the sample rate for your audio file.
```

Comment:
Why does it set the sample rate multiple times to different values? Let's have a SAMPLE_RATE const and use that everywhere

Reply:
Oh, I guess input audio is 16 kHz but output audio is 24 kHz